File: control/src/vonk_control/source_policy.py
"""Fail-closed static policy for recipe Dockerfiles and Compose documents."""

from __future__ import annotations

import json
import re
import shlex
import urllib.parse
from collections.abc import Mapping
from dataclasses import dataclass
from pathlib import PurePosixPath

import yaml

from .source_bundles import GeneratedSourceBundle

_PINNED_IMAGE = re.compile(r"^[^\s$]+@sha256:([0-9a-f]{64})$")
_NON_ROOT_USER = re.compile(r"^[1-9][0-9]*(?::[1-9][0-9]*)?$")
_COMPOSE_NAMES = frozenset(
    {"compose.yml", "compose.yaml", "docker-compose.yml", "docker-compose.yaml"}
)
_HTTPS_URL = re.compile(r"https://[^\s\"'<>]+")


@dataclass(frozen=True, slots=True)
class SourcePolicyFinding:
    code: str
    path: str
    line: int | None
    detail: str


@dataclass(frozen=True, slots=True)
class SourcePolicyReport:
    passed: bool
    source_bundle_sha256: str
    dockerfile: str
    findings: tuple[SourcePolicyFinding, ...]


class SourcePolicyError(ValueError):
    def __init__(self, report: SourcePolicyReport) -> None:
        self.report = report
        super().__init__(
            report.findings[0].detail if report.findings else "source policy failed"
        )


def inspect_build_source_policy(
    recipe: Mapping[str, object], bundle: GeneratedSourceBundle
) -> SourcePolicyReport:
    findings: list[SourcePolicyFinding] = []
    build = recipe.get("build")
    context = build.get("context") if isinstance(build, Mapping) else None
    dockerfile = build.get("dockerfile") if isinstance(build, Mapping) else None
    dockerfile_path = dockerfile if isinstance(dockerfile, str) else "Dockerfile"
    network = build.get("network") if isinstance(build, Mapping) else None
    network_mode = network.get("mode") if isinstance(network, Mapping) else None
    allowed_hosts = (
        frozenset(
            host.lower()
            for host in network.get("hosts", ())
            if isinstance(host, str)
        )
        if isinstance(network, Mapping)
        else frozenset()
    )
    expected = context.get("sha256") if isinstance(context, Mapping) else None
    if expected != bundle.sha256:
        findings.append(
            SourcePolicyFinding(
                "source.digest_mismatch",
                dockerfile_path,
                None,
                "recipe source digest does not match the canonical bundle",
            )
        )
    payload = bundle.files.get(dockerfile_path)
    if payload is None:
        findings.append(
            SourcePolicyFinding(
                "dockerfile.missing",
                dockerfile_path,
                None,
                "recipe Dockerfile is missing from the source bundle",
            )
        )
    else:
        findings.extend(
            _inspect_dockerfile(
                dockerfile_path,
                payload,
                network_mode=network_mode,
                allowed_hosts=allowed_hosts,
            )
        )
    for path, content in bundle.files.items():
        if PurePosixPath(path).name.lower() in _COMPOSE_NAMES:
            findings.extend(_inspect_compose(path, content))
    ordered = tuple(
        sorted(findings, key=lambda item: (item.path, item.line or 0, item.code))
    )
    return SourcePolicyReport(not ordered, bundle.sha256, dockerfile_path, ordered)


def enforce_build_source_policy(
    recipe: Mapping[str, object], bundle: GeneratedSourceBundle
) -> SourcePolicyReport:
    report = inspect_build_source_policy(recipe, bundle)
    if not report.passed:
        raise SourcePolicyError(report)
    return report


def dockerfile_base_images(payload: bytes) -> tuple[dict[str, str], ...]:
    """Return the ordered, unique immutable FROM authorities in a checked file."""

    text = payload.decode("utf-8")
    images: list[dict[str, str]] = []
    seen: set[str] = set()
    for _line, instruction, argument in _dockerfile_instructions(text):
        if instruction.upper() != "FROM":
            continue
        tokens = [
            token for token in argument.split() if not token.startswith("--platform=")
        ]
        reference = tokens[0] if tokens else ""
        if reference == "scratch" or reference in seen:
            continue
        matched = _PINNED_IMAGE.fullmatch(reference)
        if matched is None:
            raise ValueError("Dockerfile base image authority is invalid")
        images.append(
            {
                "manifest_digest": f"sha256:{matched.group(1)}",
                "reference": reference,
            }
        )
        seen.add(reference)
    return tuple(images)


def _inspect_dockerfile(
    path: str,
    payload: bytes,
    *,
    network_mode: object,
    allowed_hosts: frozenset[str],
) -> list[SourcePolicyFinding]:
    try:
        text = payload.decode("utf-8")
    except UnicodeDecodeError:
        return [
            _finding("dockerfile.invalid_utf8", path, None, "Dockerfile must be UTF-8")
        ]
    findings: list[SourcePolicyFinding] = []
    aliases: set[str] = set()
    final_user: tuple[str, int] | None = None
    saw_from = False
    for line_number, instruction, argument in _dockerfile_instructions(text):
        upper = instruction.upper()
        if "<<" in argument:
            findings.append(
                _finding(
                    "dockerfile.heredoc_forbidden",
                    path,
                    line_number,
                    "Dockerfile heredocs are not accepted",
                )
            )
        if upper == "FROM":
            saw_from = True
            final_user = None
            tokens = argument.split()
            tokens = [token for token in tokens if not token.startswith("--platform=")]
            base = tokens[0] if tokens else ""
            pinned = _PINNED_IMAGE.fullmatch(base)
            if pinned is not None and pinned.group(1) == "0" * 64:
                findings.append(
                    _finding(
                        "dockerfile.base_placeholder",
                        path,
                        line_number,
                        "replace the all-zero base-image placeholder with a verified linux/arm64 digest",
                    )
                )
            elif base != "scratch" and pinned is None:
                findings.append(
                    _finding(
                        "dockerfile.base_unpinned",
                        path,
                        line_number,
                        "every base image must be pinned by sha256 digest",
                    )
                )
            if len(tokens) >= 3 and tokens[-2].lower() == "as":
                aliases.add(tokens[-1])
        elif upper == "USER":
            final_user = (argument.strip(), line_number)
        elif upper == "ADD":
            findings.append(
                _finding(
                    "dockerfile.add_forbidden",
                    path,
                    line_number,
                    "ADD is forbidden; use bounded local COPY",
                )
            )
        elif upper == "ONBUILD":
            findings.append(
                _finding(
                    "dockerfile.onbuild_forbidden",
                    path,
                    line_number,
                    "ONBUILD may hide deferred build instructions",
                )
            )
        elif upper == "RUN":
            lowered = argument.lower()
            for url in _HTTPS_URL.findall(argument):
                host = urllib.parse.urlsplit(url).hostname
                if (
                    host is None
                    or network_mode != "public"
                    or host.lower() not in allowed_hosts
                ):
                    findings.append(
                        _finding(
                            "dockerfile.network_host",
                            path,
                            line_number,
                            "Dockerfile URL host is outside the declared build allowlist",
                        )
                    )
            if re.search(r"--mount\s*=\s*type\s*=\s*(?:secret|ssh)", lowered):
                findings.append(
                    _finding(
                        "dockerfile.secret_mount",
                        path,
                        line_number,
                        "secret and SSH build mounts are forbidden",
                    )
                )
            if "--network=host" in lowered or "--security=insecure" in lowered:
                findings.append(
                    _finding(
                        "dockerfile.build_privilege",
                        path,
                        line_number,
                        "host networking and insecure build execution are forbidden",
                    )
                )
        elif upper == "COPY":
            findings.extend(_inspect_copy(path, line_number, argument, aliases))
    if not saw_from:
        findings.append(
            _finding(
                "dockerfile.from_missing",
                path,
                None,
                "Dockerfile must declare a base stage",
            )
        )
    if final_user is None or _NON_ROOT_USER.fullmatch(final_user[0]) is None:
        findings.append(
            _finding(
                "dockerfile.root_user",
                path,
                final_user[1] if final_user else None,
                "the final image stage must select an explicit numeric non-root user",
            )
        )
    return findings


def _dockerfile_instructions(text: str):
    logical = ""
    start = 0
    for number, raw in enumerate(text.splitlines(), 1):
        stripped = raw.strip()
        if not logical and (not stripped or stripped.startswith("#")):
            continue
        if not logical:
            start = number
        logical += stripped[:-1].rstrip() + " " if stripped.endswith("\\") else stripped
        if stripped.endswith("\\"):
            continue
        instruction, separator, argument = logical.partition(" ")
        if instruction:
            yield start, instruction, argument.strip() if separator else ""
        logical = ""
    if logical:
        instruction, _, argument = logical.partition(" ")
        yield start, instruction, argument.strip()


def _inspect_copy(
    path: str, line: int, argument: str, aliases: set[str]
) -> list[SourcePolicyFinding]:
    findings: list[SourcePolicyFinding] = []
    try:
        tokens = (
            json.loads(argument)
            if argument.lstrip().startswith("[")
            else shlex.split(argument)
        )
    except (json.JSONDecodeError, ValueError):
        return [
            _finding(
                "dockerfile.copy_invalid",
                path,
                line,
                "COPY syntax cannot be reviewed safely",
            )
        ]
    if not isinstance(tokens, list) or not all(
        isinstance(token, str) for token in tokens
    ):
        return [
            _finding(
                "dockerfile.copy_invalid",
                path,
                line,
                "COPY syntax cannot be reviewed safely",
            )
        ]
    sources: list[str] = []
    copies_from_stage = False
    for token in tokens:
        if token.startswith("--from="):
            copies_from_stage = True
            source = token.removeprefix("--from=")
            pinned = _PINNED_IMAGE.fullmatch(source)
            if pinned is not None and pinned.group(1) == "0" * 64:
                findings.append(
                    _finding(
                        "dockerfile.copy_base_placeholder",
                        path,
                        line,
                        "replace the all-zero external COPY placeholder with a verified digest",
                    )
                )
            elif source not in aliases and pinned is None:
                findings.append(
                    _finding(
                        "dockerfile.copy_base_unpinned",
                        path,
                        line,
                        "external COPY stage must be digest-pinned",
                    )
                )
        elif not token.startswith("--"):
            sources.append(token)
    for source in sources[:-1]:
        source_path = PurePosixPath(source)
        if (
            (source.startswith("/") and not copies_from_stage)
            or source.startswith("~")
            or ".." in source_path.parts
            or "$" in source
        ):
            findings.append(
                _finding(
                    "dockerfile.copy_path",
                    path,
                    line,
                    "COPY source must stay within the canonical context",
                )
            )
    return findings


def _inspect_compose(path: str, payload: bytes) -> list[SourcePolicyFinding]:
    if len(payload) > 256 * 1024:
        return [
            _finding(
                "compose.too_large", path, None, "Compose policy input exceeds 256 KiB"
            )
        ]
    try:
        document = yaml.safe_load(payload.decode("utf-8"))
    except (UnicodeDecodeError, yaml.YAMLError):
        return [
            _finding(
                "compose.invalid",
                path,
                None,
                "Compose document is not valid UTF-8 YAML",
            )
        ]
    if not isinstance(document, Mapping) or not isinstance(
        document.get("services"), Mapping
    ):
        return [
            _finding(
                "compose.invalid",
                path,
                None,
                "Compose document must contain a services mapping",
            )
        ]
    findings: list[SourcePolicyFinding] = []
    for service_name, raw_service in document["services"].items():
        service_path = f"{path}:services.{service_name}"
        if not isinstance(raw_service, Mapping):
            findings.append(
                _finding(
                    "compose.service_invalid",
                    service_path,
                    None,
                    "Compose service must be a mapping",
                )
            )
            continue
        if raw_service.get("privileged") is True:
            findings.append(
                _finding(
                    "compose.privileged",
                    service_path,
                    None,
                    "privileged Compose services are forbidden",
                )
            )
        for key in ("network_mode", "pid", "ipc", "uts", "userns_mode"):
            if raw_service.get(key) == "host":
                findings.append(
                    _finding(
                        "compose.host_namespace",
                        service_path,
                        None,
                        f"host {key} is forbidden",
                    )
                )
        if raw_service.get("cap_add"):
            findings.append(
                _finding(
                    "compose.capabilities",
                    service_path,
                    None,
                    "added Linux capabilities are forbidden",
                )
            )
        if raw_service.get("devices"):
            findings.append(
                _finding(
                    "compose.devices",
                    service_path,
                    None,
                    "Compose device passthrough is not a build input",
                )
            )
        options = raw_service.get("security_opt", [])
        if isinstance(options, str):
            options = [options]
        if isinstance(options, list) and any(
            "unconfined" in str(item).lower() for item in options
        ):
            findings.append(
                _finding(
                    "compose.unconfined",
                    service_path,
                    None,
                    "unconfined security profiles are forbidden",
                )
            )
        findings.extend(
            _inspect_compose_volumes(service_path, raw_service.get("volumes", []))
        )
    return findings


def _inspect_compose_volumes(path: str, volumes: object) -> list[SourcePolicyFinding]:
    if not isinstance(volumes, list):
        return [
            _finding(
                "compose.volumes_invalid", path, None, "Compose volumes must be a list"
            )
        ]
    findings: list[SourcePolicyFinding] = []
    for volume in volumes:
        if isinstance(volume, str):
            source = volume.split(":", 1)[0]
            host_bind = source.startswith(("/", ".", "~"))
            socket = "docker.sock" in volume or "podman.sock" in volume
        elif isinstance(volume, Mapping):
            source = str(volume.get("source", ""))
            host_bind = volume.get("type") == "bind" or source.startswith(
                ("/", ".", "~")
            )
            socket = "docker.sock" in source or "podman.sock" in source
        else:
            findings.append(
                _finding(
                    "compose.volumes_invalid",
                    path,
                    None,
                    "Compose volume syntax is invalid",
                )
            )
            continue
        if host_bind or socket:
            findings.append(
                _finding(
                    "compose.host_bind",
                    path,
                    None,
                    "host bind mounts and container-engine sockets are forbidden",
                )
            )
    return findings


def _finding(
    code: str, path: str, line: int | None, detail: str
) -> SourcePolicyFinding:
    return SourcePolicyFinding(code, path, line, detail)

File: rust/crates/vonk-agent/src/source_policy.rs
//! Agent-side, fail-closed source policy recheck for recipe build inputs.

use std::collections::{BTreeMap, BTreeSet};

use serde::Serialize;
use serde_yaml::{Mapping, Value};

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct SourceFinding {
    pub code: &'static str,
    pub path: String,
    pub line: Option<usize>,
    pub detail: &'static str,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct SourcePolicyReport {
    pub passed: bool,
    pub dockerfile: String,
    pub findings: Vec<SourceFinding>,
}

pub fn inspect_build_source(
    files: &BTreeMap<String, Vec<u8>>,
    dockerfile: &str,
) -> SourcePolicyReport {
    let mut findings = Vec::new();
    match files.get(dockerfile) {
        Some(payload) => inspect_dockerfile(dockerfile, payload, &mut findings),
        None => findings.push(finding(
            "dockerfile.missing",
            dockerfile,
            None,
            "recipe Dockerfile is missing from the source bundle",
        )),
    }
    for (path, payload) in files {
        let name = path.rsplit('/').next().unwrap_or(path).to_ascii_lowercase();
        if matches!(
            name.as_str(),
            "compose.yml" | "compose.yaml" | "docker-compose.yml" | "docker-compose.yaml"
        ) {
            inspect_compose(path, payload, &mut findings);
        }
    }
    findings.sort_by(|left, right| {
        (&left.path, left.line.unwrap_or(0), left.code).cmp(&(
            &right.path,
            right.line.unwrap_or(0),
            right.code,
        ))
    });
    SourcePolicyReport {
        passed: findings.is_empty(),
        dockerfile: dockerfile.to_owned(),
        findings,
    }
}

pub fn dockerfile_base_images(
    files: &BTreeMap<String, Vec<u8>>,
    dockerfile: &str,
) -> Option<Vec<String>> {
    let text = std::str::from_utf8(files.get(dockerfile)?).ok()?;
    let mut seen = BTreeSet::new();
    let mut images = Vec::new();
    for (_line, instruction, argument) in dockerfile_instructions(text) {
        if !instruction.eq_ignore_ascii_case("FROM") {
            continue;
        }
        let reference = argument
            .split_ascii_whitespace()
            .find(|value| !value.starts_with("--platform="))?;
        if reference == "scratch" {
            continue;
        }
        if !pinned_image(reference) {
            return None;
        }
        if seen.insert(reference.to_owned()) {
            images.push(reference.to_owned());
        }
    }
    Some(images)
}

fn inspect_dockerfile(path: &str, payload: &[u8], findings: &mut Vec<SourceFinding>) {
    let Ok(text) = std::str::from_utf8(payload) else {
        findings.push(finding(
            "dockerfile.invalid_utf8",
            path,
            None,
            "Dockerfile must be UTF-8",
        ));
        return;
    };
    let mut aliases = BTreeSet::new();
    let mut final_user: Option<(String, usize)> = None;
    let mut saw_from = false;
    for (line, instruction, argument) in dockerfile_instructions(text) {
        if argument.contains("<<") {
            findings.push(finding(
                "dockerfile.heredoc_forbidden",
                path,
                Some(line),
                "Dockerfile heredocs are not accepted",
            ));
        }
        match instruction.to_ascii_uppercase().as_str() {
            "FROM" => {
                saw_from = true;
                final_user = None;
                let tokens = argument
                    .split_ascii_whitespace()
                    .filter(|value| !value.starts_with("--platform="))
                    .collect::<Vec<_>>();
                let base = tokens.first().copied().unwrap_or_default();
                if placeholder_image(base) {
                    findings.push(finding(
                        "dockerfile.base_placeholder",
                        path,
                        Some(line),
                        "replace the all-zero base-image placeholder with a verified linux/arm64 digest",
                    ));
                } else if base != "scratch" && !pinned_image(base) {
                    findings.push(finding(
                        "dockerfile.base_unpinned",
                        path,
                        Some(line),
                        "every base image must be pinned by sha256 digest",
                    ));
                }
                if tokens.len() >= 3 && tokens[tokens.len() - 2].eq_ignore_ascii_case("as") {
                    aliases.insert(tokens[tokens.len() - 1].to_owned());
                }
            }
            "USER" => final_user = Some((argument.trim().to_owned(), line)),
            "ADD" => findings.push(finding(
                "dockerfile.add_forbidden",
                path,
                Some(line),
                "ADD is forbidden; use bounded local COPY",
            )),
            "ONBUILD" => findings.push(finding(
                "dockerfile.onbuild_forbidden",
                path,
                Some(line),
                "ONBUILD may hide deferred build instructions",
            )),
            "RUN" => {
                let compact = argument.to_ascii_lowercase().replace(' ', "");
                if compact.contains("--mount=type=secret") || compact.contains("--mount=type=ssh") {
                    findings.push(finding(
                        "dockerfile.secret_mount",
                        path,
                        Some(line),
                        "secret and SSH build mounts are forbidden",
                    ));
                }
                if compact.contains("--network=host") || compact.contains("--security=insecure") {
                    findings.push(finding(
                        "dockerfile.build_privilege",
                        path,
                        Some(line),
                        "host networking and insecure build execution are forbidden",
                    ));
                }
            }
            "COPY" => inspect_copy(path, line, &argument, &aliases, findings),
            _ => {}
        }
    }
    if !saw_from {
        findings.push(finding(
            "dockerfile.from_missing",
            path,
            None,
            "Dockerfile must declare a base stage",
        ));
    }
    if final_user
        .as_ref()
        .is_none_or(|(user, _)| !non_root_user(user))
    {
        findings.push(finding(
            "dockerfile.root_user",
            path,
            final_user.map(|(_, line)| line),
            "the final image stage must select an explicit numeric non-root user",
        ));
    }
}

fn inspect_copy(
    path: &str,
    line: usize,
    argument: &str,
    aliases: &BTreeSet<String>,
    findings: &mut Vec<SourceFinding>,
) {
    let mut tokens = argument.split_ascii_whitespace().collect::<Vec<_>>();
    let copies_from_stage = tokens.iter().any(|value| value.starts_with("--from="));
    if let Some(from) = tokens
        .iter()
        .find_map(|value| value.strip_prefix("--from="))
        && !aliases.contains(from)
        && placeholder_image(from)
    {
        findings.push(finding(
            "dockerfile.copy_stage_placeholder",
            path,
            Some(line),
            "replace the all-zero external COPY placeholder with a verified digest",
        ));
    } else if let Some(from) = tokens
        .iter()
        .find_map(|value| value.strip_prefix("--from="))
        && !aliases.contains(from)
        && !pinned_image(from)
    {
        findings.push(finding(
            "dockerfile.copy_stage_unpinned",
            path,
            Some(line),
            "COPY --from must reference a declared stage or pinned image",
        ));
    }
    tokens.retain(|value| !value.starts_with("--"));
    let sources = if argument.trim_start().starts_with('[') {
        serde_json::from_str::<Vec<String>>(argument)
            .ok()
            .map(|items| {
                let count = items.len().saturating_sub(1);
                items.into_iter().take(count).collect()
            })
            .unwrap_or_default()
    } else {
        let count = tokens.len().saturating_sub(1);
        tokens
            .into_iter()
            .take(count)
            .map(str::to_owned)
            .collect::<Vec<_>>()
    };
    if sources.is_empty()
        || sources.iter().any(|value| {
            (value.starts_with('/') && !copies_from_stage)
                || value.split('/').any(|part| part == "..")
        })
    {
        findings.push(finding(
            "dockerfile.copy_escape",
            path,
            Some(line),
            "COPY sources must stay inside the canonical build context",
        ));
    }
}

fn inspect_compose(path: &str, payload: &[u8], findings: &mut Vec<SourceFinding>) {
    let Ok(document) = serde_yaml::from_slice::<Value>(payload) else {
        findings.push(finding(
            "compose.invalid",
            path,
            None,
            "Compose document is invalid",
        ));
        return;
    };
    let Some(services) = mapping_get(document.as_mapping(), "services").and_then(Value::as_mapping)
    else {
        findings.push(finding(
            "compose.invalid",
            path,
            None,
            "Compose document must contain a services mapping",
        ));
        return;
    };
    for service in services.values() {
        let Some(service) = service.as_mapping() else {
            findings.push(finding(
                "compose.service_invalid",
                path,
                None,
                "Compose service must be a mapping",
            ));
            continue;
        };
        if mapping_get(Some(service), "privileged").and_then(Value::as_bool) == Some(true) {
            findings.push(finding(
                "compose.privileged",
                path,
                None,
                "privileged Compose services are forbidden",
            ));
        }
        for key in ["network_mode", "pid", "ipc", "uts", "userns_mode"] {
            if mapping_get(Some(service), key).and_then(Value::as_str) == Some("host") {
                findings.push(finding(
                    "compose.host_namespace",
                    path,
                    None,
                    "host namespaces are forbidden",
                ));
            }
        }
        if nonempty_sequence(mapping_get(Some(service), "cap_add")) {
            findings.push(finding(
                "compose.capabilities",
                path,
                None,
                "added Linux capabilities are forbidden",
            ));
        }
        if nonempty_sequence(mapping_get(Some(service), "devices")) {
            findings.push(finding(
                "compose.devices",
                path,
                None,
                "Compose build metadata may not request host devices",
            ));
        }
        if sequence_strings(mapping_get(Some(service), "security_opt"))
            .iter()
            .any(|value| value.to_ascii_lowercase().contains("unconfined"))
        {
            findings.push(finding(
                "compose.unconfined",
                path,
                None,
                "unconfined security profiles are forbidden",
            ));
        }
        let volumes = mapping_get(Some(service), "volumes");
        if volumes.is_some_and(|value| !value.is_sequence()) {
            findings.push(finding(
                "compose.volumes_invalid",
                path,
                None,
                "Compose volumes must be a sequence",
            ));
        }
        for (source, explicit_bind) in volume_sources(volumes) {
            if source.contains("docker.sock") || source.contains("podman.sock") {
                findings.push(finding(
                    "compose.container_socket",
                    path,
                    None,
                    "container runtime sockets are forbidden",
                ));
            }
            if explicit_bind
                || source.starts_with('/')
                || source.starts_with("./")
                || source.starts_with("../")
                || source == "."
                || source == ".."
                || source.starts_with('~')
            {
                findings.push(finding(
                    "compose.host_bind",
                    path,
                    None,
                    "host bind mounts are forbidden",
                ));
            }
        }
    }
}

fn mapping_get<'a>(mapping: Option<&'a Mapping>, key: &str) -> Option<&'a Value> {
    mapping?.get(Value::String(key.to_owned()))
}

fn nonempty_sequence(value: Option<&Value>) -> bool {
    value
        .and_then(Value::as_sequence)
        .is_some_and(|items| !items.is_empty())
}

fn sequence_strings(value: Option<&Value>) -> Vec<&str> {
    value
        .and_then(Value::as_sequence)
        .map(|items| items.iter().filter_map(Value::as_str).collect())
        .unwrap_or_default()
}

fn volume_sources(value: Option<&Value>) -> Vec<(String, bool)> {
    value
        .and_then(Value::as_sequence)
        .map(|items| {
            items
                .iter()
                .filter_map(|item| match item {
                    Value::String(short) => short
                        .split(':')
                        .next()
                        .map(|source| (source.to_owned(), false)),
                    Value::Mapping(long) => mapping_get(Some(long), "source")
                        .and_then(Value::as_str)
                        .map(|source| {
                            let explicit_bind = mapping_get(Some(long), "type")
                                .and_then(Value::as_str)
                                == Some("bind");
                            (source.to_owned(), explicit_bind)
                        }),
                    _ => None,
                })
                .collect()
        })
        .unwrap_or_default()
}

fn dockerfile_instructions(text: &str) -> Vec<(usize, String, String)> {
    let mut result = Vec::new();
    let mut logical = String::new();
    let mut start = 0;
    for (offset, raw) in text.lines().enumerate() {
        let line = offset + 1;
        let trimmed = raw.trim();
        if logical.is_empty() && (trimmed.is_empty() || trimmed.starts_with('#')) {
            continue;
        }
        if logical.is_empty() {
            start = line;
        }
        if let Some(value) = trimmed.strip_suffix('\\') {
            logical.push_str(value.trim_end());
            logical.push(' ');
            continue;
        }
        logical.push_str(trimmed);
        let (instruction, argument) = logical
            .split_once(char::is_whitespace)
            .unwrap_or((logical.as_str(), ""));
        result.push((start, instruction.to_owned(), argument.trim().to_owned()));
        logical.clear();
    }
    if !logical.is_empty() {
        let (instruction, argument) = logical
            .split_once(char::is_whitespace)
            .unwrap_or((logical.as_str(), ""));
        result.push((start, instruction.to_owned(), argument.trim().to_owned()));
    }
    result
}

fn pinned_image(value: &str) -> bool {
    let Some((name, digest)) = value.rsplit_once("@sha256:") else {
        return false;
    };
    !name.is_empty()
        && digest.len() == 64
        && digest
            .bytes()
            .all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
}

fn placeholder_image(value: &str) -> bool {
    value
        .rsplit_once("@sha256:")
        .is_some_and(|(name, digest)| !name.is_empty() && digest == "0".repeat(64))
}

fn non_root_user(value: &str) -> bool {
    let mut parts = value.split(':');
    let valid = |part: &str| {
        !part.is_empty() && part.bytes().all(|byte| byte.is_ascii_digit()) && !part.starts_with('0')
    };
    valid(parts.next().unwrap_or_default())
        && parts.next().is_none_or(valid)
        && parts.next().is_none()
}

fn finding(
    code: &'static str,
    path: &str,
    line: Option<usize>,
    detail: &'static str,
) -> SourceFinding {
    SourceFinding {
        code,
        path: path.to_owned(),
        line,
        detail,
    }
}
